src/ast.rs

```diff
@@ -69,20 +67,40 @@ impl<'a> Program<'a> {
             .map(|index| &self.blocks[*index])
     }
 
-    fn build_block_index(blocks: &[Block]) -> Result<HashMap<u16, usize>, Error> {
+    fn build_program(mut blocks: Vec<Block>) -> Result<Program, Error> {
+        let mut program_blocks = Vec::new();
         let mut block_index = HashMap::new();
-        for (index, block) in blocks.iter().enumerate() {
+        let mut datum = Vec::new();
+        let mut index = 0;
+        for block in blocks.drain(..) {
```

Comment:
Just out of curiosity, why do you drain here?

Reply:
The idea is to take blocks, and split them into to Vec's, one containing normal blocks and other containing data blocks. For that, we can just consume the initial Vec. I guess, it is better to use `into_iter()` here, since we are not reusing the initial vector.